### PR TITLE
CCH22-13: Adjust the snippets/dependencyManagement/managingTransitiveDependencies-excludeForConfiguration sample for configuration cache

### LINKED ISSUE #21810: CCH22-13: Adjust the `snippets/dependencyManagement/managingTransitiveDependencies-excludeForConfiguration` sample for configuration cache
The sample project located under [`subprojects/docs/src/snippets/dependencyManagement/managingTransitiveDependencies-excludeForConfiguration`](https://github.com/gradle/gradle/tree/master/subprojects/docs/src/snippets/dependencyManagement/managingTransitiveDependencies-excludeForConfiguration) does not meet [the configuration cache requirements](https://docs.gradle.org/nightly/userguide/configuration_cache.html#config_cache:requirements).

The build scripts in the `groovy` and `kotlin` parts of the sample project need to be adjusted to follow the requirements.
### Context
This issue is a part of the [Gradle Configuration Cache Hackathon 2022](https://github.com/gradle/cc-hackathon-2022).

If you wish to participate and contribute the fix, please follow [the suggested workflow](https://github.com/gradle/cc-hackathon-2022/blob/main/workflow.md).
### Expected behavior

- The sample project is fixed so that the related documentation tests pass with configuration cache enabled.
- The tests related to this sample project are removed from the `testsToBeFixedForConfigurationCache` list in `subprojects/docs/build.gradle`. You can find the test IDs below.
- The doc pages that include the snippets from this sample project might need to be updated to be consistent with the sample. Feel free to ask the team for help with this! You can find the pages listed below.
### Test failure details
These tests running in `:docs:docsTest` check the sample project and are failing with configuration cache enabled:
* `snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_groovy_exclude-transitive-for-configuration.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket1SnippetsTest/snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_groovy_exclude-transitive-for-configuration.sample)
    ```
    Could not get unknown property 'configurations' for task ':printArtifacts' of type org.gradle.api.DefaultTask.
    ```

* `snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_kotlin_exclude-transitive-for-configuration.sample`[(Build Scan)](https://ge.gradle.org/s/wkszl3ueqiln2/tests/:docs:docsTest/org.gradle.docs.samples.Bucket1SnippetsTest/snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_kotlin_exclude-transitive-for-configuration.sample)
    ```
    Task `:printArtifacts` of type `org.gradle.api.DefaultTask`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
    ```

### How to reproduce
To reproduce the failure in the Gradle documentation tests, run the following in the `gradle/gradle`repository:
```sh
./gradlew :docs:docsTest \
    -PrunBrokenConfigurationCacheDocsTests=true \
    -PenableConfigurationCacheForDocsTests=true \
    --tests "*snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_groovy_exclude-transitive-for-configuration.sample" \
    --tests "*snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_kotlin_exclude-transitive-for-configuration.sample"
```
In the test build outputs, which you can find in the test report, the configuration cache problems are listed along with the URL pointing to the detailed HTML problems report that can help investigating the violation. 
You can find more instructions on working with the sample projects in the [FAQ](https://github.com/gradle/cc-hackathon-2022/blob/main/faq.md).
### Docs including the snippets from this sample project
The following doc pages include the snippets from this sample project: 
- `subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/resolution_rules.adoc`;

### PR DESCRIPTION
**Pull request by SidB3**
Fixes #21810

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/docs/src/snippets/dependencyManagement/managingTransitiveDependencies-excludeForConfiguration/kotlin/build.gradle.kts [hunk 1]
```diff
@@ -20,7 +20,8 @@ dependencies {
 // end::exclude-transitive-dependencies[]
 
 tasks.register("printArtifacts") {
+    val runtimeClasspath = configurations.getByName("runtimeClasspath")
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/docs/src/snippets/dependencyManagement/managingTransitiveDependencies-excludeForConfiguration/kotlin/build.gradle.kts:23]: When I run the tests on top of these changes, I get a failure in the test that checks the Kotlin part of the project:

* `snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_kotlin_exclude-transitive-for-configuration.sample
`

    ```
    * What went wrong:
    Configuration cache problems found in this build.
    
    1 problem was found storing the configuration cache.
    - Task `:printArtifacts` of type `org.gradle.api.DefaultTask`: cannot serialize object of type 'org.gradle.api.internal.artifacts.configurations.DefaultConfiguration', a subtype of 'org.gradle.api.artifacts.Configuration', as these are not supported with the configuration cache.
    ```

It seems that the problem comes from capturing a `Configuration` object in the task's serialized state. There is a common way to fix this by explicitly specifying the type for the variable that holds the reference to `runtimeClasspath` as `FileCollection`, [see more in the FAQ](https://github.com/gradle/cc-hackathon-2022/blob/main/faq.md#how-to-fix-cannot-serialize-gradle-script-object-references).

---

Note that after you have removed the test from the broken tests list, you need to omit `-PrunBrokenConfigurationCacheDocsTests=true` from the command to run the tests. So I ran:
```
./gradlew :docs:docsTest \
    -PenableConfigurationCacheForDocsTests=true \
    --tests "*snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_groovy_exclude-transitive-for-configuration.sample" \
    --tests "*snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_kotlin_exclude-transitive-for-configuration.sample"
```

**Review by h0tk3y — CHANGES_REQUESTED** [review]
The test `snippet-dependency-management-managing-transitive-dependencies-exclude-for-configuration_kotlin_exclude-transitive-for-configuration.sample` is still failing. Could you please fix it? See the details in the review comments.

File: subprojects/docs/src/snippets/dependencyManagement/managingTransitiveDependencies-excludeForConfiguration/kotlin/build.gradle.kts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SidB3) [reply subprojects/docs/src/snippets/dependencyManagement/managingTransitiveDependencies-excludeForConfiguration/kotlin/build.gradle.kts]: As per #21866 I modified the code to 
`val classpath: FileCollection = configurations.runtimeClasspath.get()`

**Review by nathan-contino — APPROVED** [review]
Docs LGTM

**Review by h0tk3y — APPROVED** [review]

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Pull request merged**